File: Uva12356.java
import java.io.*;
import java.util.*;
// will gives TLE , you can use cpp .
public class Uva12356
{
	public static void main(String [] args) throws IOException
	{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		PrintWriter pr = new PrintWriter(System.out);
		StringBuilder sb = new StringBuilder();
		while(true)
		{
			StringTokenizer st = new StringTokenizer(br.readLine());
			int s = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			if(s==0) break;
			int [] left = new int[s+1];
			int [] right = new int[s+1];
			for(int i = 1 ; i <= s ; i++)
			{
				left[i]=i-1;
				right[i]=i+1;
			}
			left[1] = right[s] = -1 ;
			while(b-->0)
			{
				st = new StringTokenizer(br.readLine());
				int l = Integer.parseInt(st.nextToken());
				int r = Integer.parseInt(st.nextToken());
				if(left[l]==-1)
					sb.append("* ");
				else
					sb.append(left[l]+" ");
				if(right[r]==-1)
					sb.append("*\n");
				else
					sb.append(right[r]+"\n");
				if(right[r]!=-1)
					left[right[r]] = left[l];
				if(left[l]!=-1)
					right[left[l]] = right[r];
			}
			sb.append("-\n");
		}
		pr.print(sb.toString());
		pr.close();
		br.close();
	}
}

File: uva12356.cpp
#include <cstdio>
using namespace std ;
int left[(int)1e5+5] , right[(int)1e5+5];
int main()
{
	int s , b , L , R ;
	while(scanf("%d%d",&s,&b) , s )
	{
		for(int i  = 1 ; i <= s ;++i)
		{
			left[i] = i-1 ;
			right[i] = i+1 ;
		}	
		left[1] =right[s] =  -1 ;
		while(b-->0)
		{
			scanf("%d%d",&L,&R);
			
			if(left[L]!=-1)
				printf("%d ",left[L]);
			else
				printf("* ");

			if(right[R]!=-1)
				printf("%d\n",right[R]);
			else
				printf("*\n");
			
			if(right[R]!=-1)
				left[right[R]] = left[L];
			if(left[L]!=-1)
				right[left[L]] = right[R];
		}
		printf("-\n");
	}
	return 0 ;
}
